### ISSUE TITLE
[question] Resolving missing binaries problem if python-require package changed in one of dependencies

### ISSUE THREAD

**Issue by apyvovar** [conan-io/conan#15327]
### What is your question?

Let's assume, we have `MyApplication` with the following requirements:

```
    def requirements(self):
        self.requires("some_package/1.2.3")
        self.requires("pkg_b/1.2.3")
        self.requires("pkg_c/1.2.3")
```
`pkg_b` itself:
```
    def requirements(self):
        self.requires("pkg_c/[>=1.0.0 <2.0.0]")
```
`pkg_c` itself:
```
    def requirements(self):
        self.requires("some_package/[>=1.0.0 <2.0.0]")
```
Everything works here, and now `some_package/1.2.4` is introdued with no changes, but with new minor version of `python-require`.
Now, `MyApplication` suddenly is broken, because `pkg_b` binaries are now `Missing`, according to conan (of course, in the release day :) ). It's now required to rebuild the whole depndency tree.

As far as I can see, this is expected behavior, because of change of in `python-require` package. Is that right?

I actually have 2 questions here:
1. If `MyApplication` explicitly requests the older version of `some_package`, why rebuild of `pkg_b` and `pkg_c` is required in this case? We are not using newer `some_package` here, so I would expect no rebuild request here. Instead, with no changes in conanfile, and with explicitly set valid and compatible dependencies, we still have our whole dependency tree broken.

2. How to change default packge ID mode for `python-require`?
We have dozens of packages now in our private repository and changing global config would break all our recipes and applications, which relies on conan (all of them use `python-require`). 
So, I'm looking for easier way to pull this off, maybe to change packages one by one with using conanfile attiribute, but there's no `package_id_python_mode`, as, for example `package_id_embed_mode`

UPD:
3. I'm thinking of changing default package ID mode, as documented. Is there any way to do it without breaking all existing recipes?

Thank you in advance.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @apyvovar 

Thanks for your question

> If MyPackage explicitly requests the older version of some_package, why rebuild of pkg_b and pkg_c is required in this case? We are not using newer some_package here, so I would expect no rebuild request here. Instead, with no changes in conanfile, and with explicitly set valid and compatible dependencies, we still have our whole dependency tree broken.

It is not very clear what is ``MyPackage``. You have defined ``MyApplication`` and ``some_package``, but not ``MyPackage``.

> Everything works here, and now some_package/1.2.4 is introdued with no changes, but with new minor version of python-require.

It is not clear how is it introduced. Changing the ``requires`` in ``MyApplication``? Otherwise, just creating it doens't seem that it will result in resolving to 1.2.4, because it is defining a pinned dependency to 1.2.3. It is also not clear if this is the only package that contains a python-require, it would be helpful to have the definition of ``some_package`` too.



> 3. I'm thinking of changing default package ID mode, as documented. Is there any way to do it without breaking all existing recipes?

Changing the package-id-modes do not break any recipe. It will only result in new computed ``package_ids`` and it will require building from source those new binaries, but that shouldn't require any change in recipes, so recipes shouldn't be broken for this change.

> How to change default packge ID mode for python-require?

It is not clear why you are trying to change it or what problem you are trying to solve. The default mode is pretty useful, allowing full control over when things have to be built:
- increasing the patch means binaries don't need to be rebuilt
- increasing the minor is done when we want to force the binaries to be rebuilt
- increasing the major is done when we are doing breaking API changes, so the major will typically be discarded, because the range will be like ``python_requires = "mypyrequire/[>=1.0 <2]"``

So increasing the minor effectively means "I want the packages using this python-require to rebuild their binaries". You can change the python-require-id-mode, but that basically means shifting the above logic left or right in the version digits, but still the same logic.

> If MyPackage explicitly requests the older version of some_package, why rebuild of pkg_b and pkg_c is required in this case? 

We need to check this. If the packages are static-libraries requiring other static libraries, they won't require re-building unless the minor version of the dependencies change. Having them changing only their python-requires should not be requiring the consumers to re-build (only applications, but not static libraries). So a full reproducible case would be necessary to validate this.

**Comment by apyvovar** [comment]
> It is not very clear what is MyPackage. You have defined MyApplication and some_package, but not MyPackage.

Oh, sorry. It's the same. Basically, it means "root package consumer". I've updated the original message.

> It is not clear how is it introduced.

What I meant here is new version of `some_package` was created. Something like this:
| Before  | After |
| ------------- | ------------- |
| name = "some_package"  | name = "some_package" |
| version = "1.2.3"  | version = "1.2.4" |
| python_requires = "conan_utils/2.1.0"  | python_requires = "conan_utils/2.2.0" |

So, `some_package` was rebuilt with updated `conan_utils` python_requires package, minor version of which was bumped.

> It is not clear why you are trying to change it or what problem you are trying to solve

After `some_package/1.2.4` is released, dependencies of `MyApplication` binaries of `pkg_b` and `pkg_c` became `Missing` and conan fails with error. That's a big problem, because we don't uploading sources to remote with packages, so each and every dependency has to be rebuilt and uploaded to Artifactory one by one.

I want to emphasize here, that `MyApplication` explicitly set `some_package/1.2.3` - old version, which worked before. I just simply don't understand in what way `some_package/1.2.4` is somehow related here.

> increasing the minor is done when we want to force the binaries to be rebuilt

No, we don't want to force the binaries to be rebuilt. Increasing the minor version is "add functionality in a backward compatible manner", according to semver. When we add functionality to python script, we don't need anything to rebuilt.

> So a full reproducible case would be necessary to validate this.

Ok, I'll try to create reproducible example

**Comment by memsharded** [comment]
> No, we don't want to force the binaries to be rebuilt. Increasing the minor version is "add functionality in a backward compatible manner", according to semver. When we add functionality to python script, we don't need anything to rebuilt.

Semver is more or less ok for API-compatibility. But Conan (or C++ in general) also needs to model ABI, and the need of re-building packages even when the API is fully compatible. There are multiple cases of this, like when modifying headers that contains templates or structs definitions that can change memory layout, or when a static library is linked by a shared library.

So it is necessary to define the different cases:
- Breaking API => Major
- Not breaking API but need to rebuild binaries => Minor
- Not breaking API, no need to rebuild binaries => Patch

There are cases that the API doesn't change at all, and still binaries must be rebuild, like the shared->static case. Conan will automatically manage this, even if the version is not bumped, as long as ``package_type`` is correctly defined.

In short, semver is not optimal for C/C++ build and compilation model, better take the above considerations in mind when defining the versioning scheme

**Comment by memsharded** [comment]
> After some_package/1.2.4 is released, dependencies of MyApplication binaries of pkg_b and pkg_c became Missing and conan fails with error. 

This should only happen if ``pkg_b`` and ``pkg_c`` are shared libraries linking a static or header-only library or if ``some-package`` is a header-only library, and that is totally expected and desired behavior, the risk is too high, things will be mosly broken if things are not rebuild. If all are static libraries, then it shouldn't happen, as re-building is not necessary.

**Comment by apyvovar** [comment]
> This should only happen if pkg_b and pkg_c are shared libraries linking a static or header-only library or if some-package is a header-only library, and that is totally expected and desired behavior, the risk is too high, things will be mosly broken if things are not rebuild. If all are static libraries, then it shouldn't happen, as re-building is not necessary.

That is understandable in case, I'm actually using version 1.2.4. But I am not. In my case, all dependencies - `some_package`, `pkg_b` and `pkg_c` - are requested explicitly in `MyApplication` with exact versions. This means the binaries and dependency graph must remain absolutely the same regaredless of versions of other released depedencies on Artifactory. 

Maybe I don't fully understand how it meant to work, but in our development flow the release of new package doesn't mean, that all packages and applications in organisation should immediately use it. Changes have to be tested first and applied only if everything works as expected

I'm sorry, but I still don't understand why rebuild is required here.

**Comment by apyvovar** [comment]
Let me also rephrase my other question. If I want to set package ID mode for some package (not default for all in config), I have only the following options:
```
package_id_embed_mode = ...
package_id_non_embed_mode = ...
package_id_unknown_mode = ...
```
What should I do if I want to change package ID mode for `python_require` package in the same manner? 
Looks like, the only possibility is via config file, but I don't want change default for all.
What I'd like to see is possibility to do something like this in conanfile attribute:
```
package_id_python_mode = ...
```
Exactly like the entry in config file, but for single package.
Is this possible?

**Comment by memsharded** [comment]
> That is understandable in case, I'm actually using version 1.2.4. But I am not. In my case, all dependencies - some_package, pkg_b and pkg_c - are requested explicitly in MyApplication with exact versions. This means the binaries and dependency graph must remain absolutely the same regaredless of versions of other released depedencies on Artifactory.

There are different aspects here. Depending on a fixed version doesn't guarantee anything if that version for example gets a new revision or a new binary. For example if ``pkg_b`` and ``pkg_c`` are static libraries, even if they are the same version, if the are re-built from source, because they depend on a transitive dependency, lets say a new header-only new version, both ``pkg_b`` and ``pkg_c`` will be different binaries with different functionality that has been compiled from the headers of the transitive dependency. And if that is the case, of course the final consumer application will have to be re-built, to link both ``pkg_b`` and ``pkg_c`` new binaries, of course, otherwise the application will not integrate the changes done in the upstream code.

In short, when dependencies change, in many situations it is necessary to re-build the consumers binaries.
In your described situation, there is not enough information to know why it is happening, because there are many other factors that are important, settings, options, package-types, configuration, etc. If we want to understand what is happening we would need a complete reproducible case, that we can run and check, otherwise it is nearly impossible to diagnose.

> Maybe I don't fully understand how it meant to work, but in our development flow the release of new package doesn't mean, that all packages and applications in organisation should immediately use it. Changes have to be tested first and applied only if everything works as expected

The way new changes are integrated in the downstream is defined by the versioning scheme. You can have version ranges that exclude certain versions, if you are releasing a new version, in the same server repository together with the other packages that contain ``requires`` with version-ranges that include the new version, then the new version will automatically be used, this is how dependency resolution works. You can follow different strategies:

- Publish a new version to a secondary server repository. From there run the tests, build necessary packages of consumers, etc, and only copy the new package version and new binaries to the main server repository when everything is working and approved. This is a very recommended CI flow
- Make sure that the new version is outside of the defined version ranges. You can use only exact dependencies and not version-ranges, that way, when a new version is published it will not be automatically used. Then, when packages want to use the newer version, change their requirements and publish new versions of those too. This can be a bit tedious and require a lot of manual work, but is "exact" with versions
- Use lockfiles to guarantee you are using a fixed set of versions irrespective of the versions that have been published to the server.

**Comment by apyvovar** [comment]
Ok, thank you for all the info, I'll try to adjust my development flow to better fit this logic.

And, please, take a look at this https://github.com/conan-io/conan/issues/15327#issuecomment-1867146009 , it would be helpful for some of my python-require packages

**Comment by memsharded** [comment]
I am following up on your comment above, about a potential ``package_id_python_mode``.

Actually I have already some code to implement it. But then I recalled the reason why this was not provided initially:

- The C and C++ libraries can have different versioning schemes, depends on the organization versioning policies, specific implementation details of the libraries, etc. It makes sense to be able to change the default behaviors.
- Conan ``python_requires`` is totally a Conan specific invention. It defines a ``package_id`` policy that makes sense to cover all potential cases that users might want to cover, those are mostly 3:
  - The changes in the ``python_requires`` do not require consumers to re-build from sources, not causing new ``package_id``. These changes can be represented by bumping the "patch" version of the ``python_requires``
  - The changes in the ``python_requires`` require consumers to re-build from sources. Bump the "minor" version of the ``python_require``
  - The changes in the ``python_requires`` would be breaking, requiring the consumers to change their own python code of their conanfiles. Bump the "major" version of the ``python_requires``

Any other potential mode would result in just shifting left or right this logic. But why changing the default logic, if this is Conan specific? It would be better if the default ``minor_mode`` is followed, as it is already the default, isn't it? What other mode would you like to select for your python requires and why?

**Comment by apyvovar** [comment]
Okay, I see the logic behind it.

On the contrary, in our organisation we believe, that increasing version of any Conan package must represent the changes behind it (according to semver), and not the `package_id` policy exclusively.

E.g.: bugfix should be introduced with increased patch version of package, while adding new python function should lead to increasing minor version. And in both cases, it should not require the consumers to re-build from sources. 

For us it's important, because, as I've previously written, we're distributing binaries only, so if package ID changes, users have to wait for all dependency tree to rebuild on CI and the whole organisation basically stops, because of `missing binary` error.

And so, while I'm adding lots of new APIs to my python-require package, I'm forced to increase patch version without any other choice. Minor bug fix and massive addition to API is equal changes according to package version. This makes our code history look strange IMO.

Moreover, as I can see, this actually can be done in config file
```
core.package_id:default_python_mode
```

And I'm confused why is it allowed in config and shouldn't be possible in conanfile itself in the same way as other modes.

**Comment by memsharded** [comment]
> E.g.: bugfix should be introduced with increased patch version of package, while adding new python function should lead to increasing minor version. And in both cases, it should not require the consumers to re-build from sources.

> For us it's important, because, as I've previously written, we're distributing binaries only, so if package ID changes, users have to wait for all dependency tree to rebuild on CI and the whole organisation basically stops, because of missing binary error.

This is important, while I understand the wish to use ``semver``, it is important to highlight that it doesn't work for C/C++, you might want to use one of the versions digits to represent changes in the API even if they don't require re-building the consumers, but if you are missing a digit that allows you to represent the case that the consumers need to be re-built, independently of the API changes, then you will be blocked in many situations, or even worse, you will have production bugs. Some examples:

- A python requires that contains a ``generate()`` method and in the generate method you realize that you forgot a critical compile c++ flag, that renders your binaries insecure. So you add in your ``python_require`` a ``toolchain.extra_cxxflags = ...``. This change didn't change the API of the python-requires at all, yet it needs to induce a re-build of the consumers that depend on that python-require, or they won't use the new cxxflags and will be deployed to production without them
- A normal C++ library requirement, gets changes in the implementation of some templated functions in the headers, or alters the order of attributes in a struct. In both changes there is no change in the API at all, but there are changes in the ABI. If you can't define a way that indicates the consumers that they have to be re-built from sources, you will likely have runtime binary incompatibility issues or running all code in production

If you use the "minor" to represent API changes that do not require a re-build of the consumers, you will run out of digits to use, because the major is still needed to represent breaking changes that will need changing the code of the consumers and those are the changes excluded in version ranges such as ``[>=1.0 <2]``

> And so, while I'm adding lots of new APIs to my python-require package, I'm forced to increase patch version without any other choice. Minor bug fix and massive addition to API is equal changes according to package version. This makes our code history look strange IMO.

You might want to create a more elaborate versioning scheme using the micro versions:
- Micro: Bugfixes, not necessary to rebuild consumers
- Patch: API changes that are fully ABI compatible, not necessary to rebuild consumers
- Minor: ABI changes, necessary to rebuild consumers
- Major: Breaking API changes

But beware that if you select your versioning scheme with only 3 digits and representing the traditional semver definition, you might have a history that represents the API changes according to that, but the risk of having binary compatibility issues, runtime errors or production binaries running old code that doesn't include the latest changes is very high, because the information "something needs to be rebuilt" is being discarded in favor of that traditional semver definition and history.

> For us it's important, because, as I've previously written, we're distributing binaries only, so if package ID changes, users have to wait for all dependency tree to rebuild on CI and the whole organisation basically stops, because of missing binary error.

Well, if the versioning is used, ``missing error`` happens to protect you against those binaries errors. When the package-id-modes and the versioning scheme is correctly used, the ``missing binary error`` is not an error in the sense that something failed, but yet, until those new binaries are built from source, your code changes are not in production at all, and your binaries will run with legacy and buggy code, potentially with runtime memory issues. And you will not know what things actually need to be re-built to solve that, because you removed that information from the versioning scheme.

> core.package_id:default_python_mode
> And I'm confused why is it allowed in config and shouldn't be possible in conanfile itself in the same way as other modes.

Because existing libraries might have different versioning schemes, depending if they use templates or not, if they use ``pimpl`` idioms, if they are C or C++, because the binary compatibility and compilation embedding and linkage mode can vary. But having your own Conan ``python-requires`` to use different package-id modes sounds like an unnecessary and self-inflicted variability. Having the whole organization understand what a version bump to any python-require is an advantage, and there shouldn't be any reason to have a variable one in different python-requires.

**Comment by apyvovar** [comment]
Thank you for such detailed post, I see the logic behind it.

One more thing, which still bothers me, is that changing default python mode in config file (I case, we decide to do it) would in fact require rebuilt of absolutely all existing packages at once. As I've mentioned, rebuild is quite challenging for our development flow, and so, `package_id_python_mode` would be very nice tool to do it gradually, one by one and change the default later, when all packages are already using it.

Is there a chance to introduce it with note, that its' usage meant to be temporary or for some unique cases like mine (gradual transition of existing packages to new package id python mode)?

**Comment by memsharded** [comment]
> One more thing, which still bothers me, is that changing default python mode in config file (I case, we decide to do it) would in fact require rebuilt of absolutely all existing packages at once. As I've mentioned, rebuild is quite challenging for our development flow, and so, package_id_python_mode would be very nice tool to do it gradually, one by one and change the default later, when all packages are already using it.

That would be a valid reason, I'll check it with the team and consider adding the per-recipe mode.

I'd also suggest try to plan to improve the automation to build missing binaries. We will be working this 2024 year in improving our own ConanCenter CI, and we will be releasing tools, documentation, examples on CI flows that hopefully help users to automate the creation of binaries, so when something changes in a package, your needed binaries are automatically built. Please keep tuned to our releases and announcements, we will be sharing progress. Thanks!

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/15453 has been merged, it will be available next Conan 2.1, thanks for the feedback!
